`datetime(number, mode)`

ericnormand opened the pull request:
Closes QUE-1504

ericnormand committed:
Empty

ericnormand committed:
Merge branch 'master' into datetime-number

ericnormand committed:
No need for intersection in test

ericnormand committed:
Correct modestring casing

ericnormand committed:
Test datetime(number, mode) for s, ms, us, ns

ericnormand committed:
Allow uppercase and hyphens but normalize them.

ericnormand committed:
Don't allow hyphens in frontend.

ericnormand committed:
Implement unix casts for mongo

ericnormand committed:
Convert floats to long

ericnormand committed:
Use hyphenated keywords internally

ericnormand committed:
expand docstring

ericnormand committed:
Try other implementation of nanoseconds for BigQuery

ericnormand committed:
Handle sqlserver unix timestamps with multimethod

ericnormand committed:
More flexible date type handling

ericnormand committed:
Use integer division with bigquery

ericnormand committed:
Merge branch 'master' into datetime-number

ericnormand committed:
datetime(number) frontend tests

ericnormand committed:
Undo change to package.json

romeovs commented on frontend/src/metabase-lib/v1/expressions/config.ts:
```diff
@@ -326,23 +326,41 @@ const CONVERSION = defineClauses(
           name: t`mode`,
           type: "string",
-          description: t`The mode indicating the format. One of: "simple", "iso", "simplebytes", "simpleiso". Default is "iso".`,
+          description: t`The mode indicating the format. One of: "simple", "iso", "simpleBytes", "isoBytes", "unixSeconds", "unixMilliseconds", "unixMicroseconds", "unixNanoseconds". Default is "iso".`,
```
Please mark code-like stuff as such using <code>\\`</code>
```suggestion
          description: t`The mode indicating the format. One of: \`"simple"\`, \`"iso"\`, \`"simpleBytes"\`, \`"isoBytes"\`, \`"unixSeconds"\`, \`"unixMilliseconds"\`, \`"unixMicroseconds"\`, \`"unixNanoseconds"\`. Default is \`"iso"\`.`,
```

romeovs commented:
FE looks good! Please also get BE review.

ericnormand committed:
Wrap code-like stuff in ``

metamben commented on src/metabase/lib/schema/expression/temporal.cljc:
```diff
@@ -60,20 +62,41 @@
+
+   Is lenient on case and hyphens."
+  [s]
+  (get datetime-mode-map (datetime-mode->string s)))
```
I wonder if it's worth special casing already normalized modes.

metamben commented on test/metabase/query_processor_test/cast_test.clj:
```diff
@@ -986,3 +998,48 @@
+  [_]
+  "2025-07-02 18:33:35")
+
+;; sqlserver's sql.qp/unix-timestamp->honeysql rounds to minutes
```
```suggestion
;; sqlserver's sql.qp/unix-timestamp->honeysql truncates to minutes
```

metamben commented on test/metabase/query_processor_test/cast_test.clj:
```diff
@@ -986,3 +998,48 @@
+                          (lib/with-fields [(lib.metadata/field mp (mt/id :orders :id))])
+                          (lib/expression "date_number" (lib/+ 0 (* seconds-timestamp multiple)))
+                          (as-> q
+                                (let [column (->> q lib/visible-columns (filter #(= "date_number" (u/lower-case-en (:name %)))) first)]
```
Could use `m/find-first` instead of `(comp first filter)`.

ericnormand committed:
Better comment

ericnormand committed:
use m/find-first

ericnormand committed:
Merge branch 'master' into datetime-number

ranquild reviewed (changes requested):
Several issues with how `mode` is handled:

1. When using a `mode` with 2 words, saving and re-opening an expression leads to an error
<img width="583" alt="Screenshot 2025-07-07 at 12 28 57" src="https://github.com/user-attachments/assets/6f040f9c-16e1-4041-8768-35aeba93a809" />

2. No "Type" errors when using something wrong instead of mode. Something is different with the schema for `datetime` compared to other schemas. Compare:
<img width="568" alt="Screenshot 2025-07-07 at 12 26 30" src="https://github.com/user-attachments/assets/48d83638-ff7b-4104-bfbb-6728a6906525" />
<img width="769" alt="Screenshot 2025-07-07 at 12 31 54" src="https://github.com/user-attachments/assets/0f90c843-e2b9-463e-9f16-6224518394ac" />

3. No type checking for cases like `datetime(10, "iso")`. I expect that this is something that should be done on the schema level, e.g. with a dispatch based on `mode`:
<img width="806" alt="Screenshot 2025-07-07 at 12 32 49" src="https://github.com/user-attachments/assets/bed2940f-0e31-412e-8a07-6fec713d4f55" />
<img width="1037" alt="Screenshot 2025-07-07 at 12 32 46" src="https://github.com/user-attachments/assets/7985bc91-1369-4450-992c-4a3f66e66030" />

ericnormand commented:
> Several issues with how `mode` is handled:
> 
> 1. When using a `mode` with 2 words, saving and re-opening an expression leads to an error
> 

I made it not support hyphenated modes in the frontend. The hyphenated version is only internally available as a keyword.

> 2. No "Type" errors when using something wrong instead of mode. Something is different with the schema for `datetime` compared to other schemas. Compare:
> 

I would need @romeovs 's help to answer that one.

> 3. No type checking for cases like `datetime(10, "iso")`. I expect that this is something that should be done on the schema level, e.g. with a dispatch based on `mode`:
> 

I can look into this.

ericnormand committed:
Add a line to changelog

ericnormand committed:
Handle different types better in lib schema

ericnormand commented:
@ranquild I've updated the lib schema to reflect the three types we support (string, number, binary) and removed the frontend validator. I think it addresses your concerns. Would you look at it again?

ericnormand commented:
@ranquild Geez, maybe not. It's still saying "Invalid expression" instead of "Types are incompatible".

ericnormand commented:
> @ranquild Geez, maybe not. It's still saying "Invalid expression" instead of "Types are incompatible".

Now I'm thinking it's because the schema is correct but more complex.

romeovs committed:
Add special case for mode in formatter

romeovs committed:
Add mode to expression options

romeovs committed:
Add special case for mode in parser

ericnormand committed:
Error message

ericnormand committed:
Schema with options

ericnormand committed:
Add options map to legacy schema for datetime

romeovs committed:
Remove comment

romeovs committed:
Add unit test for mode option in compiler

romeovs committed:
Add test for formatting the mode option in datetime

ericnormand committed:
Better round-tripping for :datetime function

romeovs committed:
Add :mode to allowed keys for options marshalling

ericnormand committed:
Fix expression to add mode to options map

ericnormand committed:
Test datetime() legacy/lib roundtripping

ericnormand committed:
Fix normalization

ericnormand committed:
Remove println

ericnormand committed:
Mongo datetime() with options map

ericnormand committed:
Merge branch 'master' into datetime-number

ericnormand committed:
Merge branch 'master' into datetime-number

ranquild commented on frontend/src/metabase-lib/v1/expressions/pratt/compiler.ts:
```diff
@@ -266,6 +270,35 @@ function compileFunctionCall(node: Node, ctx: Context): Lib.ExpressionParts {
+    return "iso";
+  }
+
+  check(
```
This is redundant and seems hacky. Just `isStringLiteral` is enough

ranquild reviewed (approved):
@romeovs small comment

romeovs committed:
Remove duplicate check

ericnormand committed:
Merge branch 'datetime-number' of github.com:metabase/metabase into datetime-number

ericnormand committed:
Merge branch 'master' into datetime-number

ericnormand committed:
Add simple bytes to compiler
